machineconfiguration/v1alpha1/types_machineconfignode.go

```diff
@@ -0,0 +1,169 @@
+type MachineConfigNodeStatus struct {
+	// conditions represent the observations of a machine config node's current state.
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+	// observedGeneration represents the generation observed by the controller.
+	// This field is updated when the controller observes a change to the desiredConfig in the configVersion of the machine config node spec.
+	// +kubebuilder:validation:Required
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+	// configVersion describes the current and desired machine config for this node.
+	// The current version represents the current machine config for the node and is updated after a successful update.
+	// The desired version represents the machine config the node will attempt to update to. This gets set before the machine config operator validates the new
+	// machine config against the current machine config.
+	// This desiredConfig has been validated in the machine config operator as one that is valid and that exists.
+	// +kubebuilder:validation:Required
+	ConfigVersion MachineConfigNodeStatusMachineConfigVersion `json:"configVersion"`
+}
+
+// MachineConfigNodeStatusMachineConfigVersion holds the current and desired config versions as last updated in the MCN status.
+// When the current and desired versions are not matched, the machine config pool is processing an upgrade and the machine config node will
+// monitor the upgrade process.
+// When the current and desired versions are not match, the machine config node will ignore these events given that certain operations happen both during the MCO's upgrade mode and the daily operations mode.
+type MachineConfigNodeStatusMachineConfigVersion struct {
+	// current is the name of the machine config currently in use on the node.
+	// This value is updated once the machine config daemon has completed the update of the configuration for the node.
+	// This value should match the desired version unless an upgrade is in progress.
+	// Must be a lowercase RFC-1123 hostname (https://tools.ietf.org/html/rfc1123)
+	// It may consist of only alphanumeric characters, hyphens (-) and periods (.)
+	// and must be at most 253 characters in length.
+	// +kubebuilder:validation:MaxLength=253
+	// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])(\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9]))*$`
+	// +optional
+	Current string `json:"current"`
+	// desired is the MachineConfig the node wants to upgrade to.
+	// This value gets set in the machine config node spec once the machine config has been validated.
+	// This value gets set in the machine config node status once the machine config has been targeted for upgrade and before it is validated.
```

Comment:
This seems like it's the wrong way around to me.

The flow I'd expect is:
* machine config controller generates a new rendered config
* <something> assigns this to the MCP
* <something> updates the Node annotation to match the new desired MCP rendered config
* Above <something> sets the spec of the desired config on the MCN
* MCD (?) observes a difference between spec and status desired versions
* MCD (?) validates that there is a viable transition between `status.current` and `spec.desired`
* MCD (?) updates the observedGeneration and `status.desired` once it has validated the viable transition

Does that sound right to you? Can you fill in the gaps/correct me on which components do what?

Reply:
sorry you are right, flipped these around and they are fixed.